crates/nu-cli/src/repl.rs

```diff
@@ -45,18 +46,51 @@ const PRE_EXECUTE_MARKER: &str = "\x1b]133;C\x1b\\";
 const RESET_APPLICATION_MODE: &str = "\x1b[?1l";
 
 ///
-/// The main REPL loop, including spinning up the prompt itself.
+/// This shared stack exists to make sure that we never explicitly copy the stack.
 ///
+trait SharedStack {
+    fn convert_env_values(&mut self, engine_state: &mut EngineState) -> Option<ShellError>;
+    fn add_env_var(&mut self, var: String, value: Value);
+    fn get_env_var(&mut self, envine_state: &EngineState, var: &str) -> Option<Value>;
+    fn child_stack(&self) -> Stack;
+    fn get_unique_mut(&mut self) -> &mut Stack;
+}
+
+impl SharedStack for Arc<Stack> {
+    fn get_unique_mut(&mut self) -> &mut Stack {
+        Arc::get_mut(self).expect("More than one reference to the stack found!")
+    }
+    fn convert_env_values(&mut self, engine_state: &mut EngineState) -> Option<ShellError> {
+        let stack = self.get_unique_mut();
+        convert_env_values(engine_state, stack)
+    }
+
+    fn add_env_var(&mut self, var: String, value: Value) {
+        let stack = self.get_unique_mut();
+        stack.add_env_var(var, value);
+    }
+
+    fn get_env_var(&mut self, engine_state: &EngineState, var: &str) -> Option<Value> {
+        let stack = self.get_unique_mut();
+        stack.get_env_var(engine_state, var)
+    }
+    fn child_stack(&self) -> Stack {
+        Stack::with_parent(self.clone())
+    }
+}
+
+/// The main REPL loop, including spinning up the prompt itself.
 pub fn evaluate_repl(
     engine_state: &mut EngineState,
-    stack: &mut Stack,
+    stack: Stack,
     nushell_path: &str,
     prerun_command: Option<Spanned<String>>,
     load_std_lib: Option<Spanned<String>>,
     entire_start_time: Instant,
 ) -> Result<()> {
     use nu_cmd_base::hook;
     use reedline::Signal;
+    let mut stack = Arc::new(stack);
```

Comment:
```suggestion
    let mut unique_stack = stack;
```
You could create a variable here to hold a unique, owned Stack. It should be fine to use this until the Arc creation below. (Feel free to name the variable whatever you think is best.)